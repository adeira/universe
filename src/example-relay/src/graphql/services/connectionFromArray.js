// @flow strict-local

import {
  cursorToOffset,
  connectionFromArray as relayConnectionFromArray,
  offsetToCursor,
  type Connection,
  type ConnectionArguments,
} from '@adeira/graphql-relay';

type Args = {
  ...ConnectionArguments,
  ...
};
// There is open discusion how to handle bidirectional pagination:
// https://github.com/graphql/graphql-relay-js/issues/58
//
// Until Relay devs decide, let's try our implementation
export default function connectionFromArray<T>(data: $ReadOnlyArray<T>, args: Args): Connection<T> {
  const { edges, pageInfo } = relayConnectionFromArray(data, args);
  const firstCursor = offsetToCursor(0);
  const lastCursor = offsetToCursor(data.length - 1);
  const lowerBound = cursorToOffset(pageInfo.startCursor ?? firstCursor);
  const upperBound = cursorToOffset(pageInfo.endCursor ?? lastCursor);

  // there are some data "before" current slice
  if (lowerBound > 0) {
    pageInfo.hasPreviousPage = true;
  }
  // there are some data "after" current slice
  if (upperBound < data.length - 1) {
    pageInfo.hasNextPage = true;
  }

  return { edges, pageInfo };
}
