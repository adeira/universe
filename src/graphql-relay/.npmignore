__tests__
test-helpers
BUILD.bazel
BUILD
