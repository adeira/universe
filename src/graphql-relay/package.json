{
  "name": "@adeira/graphql-relay",
  "private": false,
  "sideEffects": false,
  "version": "0.3.0",
  "description": "A library to help construct a graphql-js server supporting react-relay.",
  "license": "MIT",
  "homepage": "https://github.com/adeira/universe/tree/master/src/graphql-relay",
  "bugs": {
    "url": "https://github.com/adeira/universe/issues"
  },
  "repository": {
    "type": "git",
    "url": "http://github.com/adeira/universe.git"
  },
  "main": "src/index",
  "peerDependencies": {
    "graphql": "^15.0.0"
  },
  "dependencies": {
    "@babel/runtime": "^7.13.17"
  }
}
