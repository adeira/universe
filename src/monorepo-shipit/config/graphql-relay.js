// @flow strict

import type { ConfigType } from '../ConfigType.flow';

module.exports = ({
  getStaticConfig() {
    return {
      repository: 'git@github.com:adeira/graphql-relay.git',
    };
  },
  getPathMappings() {
    return new Map([['src/graphql-relay/', '']]);
  },
}: ConfigType);
