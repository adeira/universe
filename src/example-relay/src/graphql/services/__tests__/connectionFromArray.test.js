// @flow

import { offsetToCursor } from 'graphql-relay';

import connectionFromArray from '../connectionFromArray';

// Tests of connectionFromArray are copied from origin implementation:
// https://github.com/graphql/graphql-relay-js/blob/master/src/connection/__tests__/arrayconnection.js
const letters = ['A', 'B', 'C', 'D', 'E'];

it('returns all elements without filters', () => {
  expect(connectionFromArray(letters, {})).toMatchSnapshot();
});

it('respects a smaller first', () => {
  expect(connectionFromArray(letters, { first: 2 })).toMatchSnapshot();
});

it('respects an overly large first', () => {
  expect(connectionFromArray(letters, { first: 10 })).toMatchSnapshot();
});

it('respects a smaller last', () => {
  expect(connectionFromArray(letters, { last: 2 })).toMatchSnapshot();
});

it('respects an overly large last', () => {
  expect(connectionFromArray(letters, { last: 10 })).toMatchSnapshot();
});

it('respects first and after', () => {
  expect(
    connectionFromArray(letters, {
      first: 2,
      after: offsetToCursor(1),
    }),
  ).toMatchSnapshot();
});

it('respects first and after with long first', () => {
  expect(
    connectionFromArray(letters, {
      first: 10,
      after: offsetToCursor(1),
    }),
  ).toMatchSnapshot();
});

it('respects last and before', () => {
  expect(
    connectionFromArray(letters, {
      last: 2,
      before: offsetToCursor(3),
    }),
  ).toMatchSnapshot();
});

it('respects last and before with long last', () => {
  expect(
    connectionFromArray(letters, {
      last: 10,
      before: offsetToCursor(3),
    }),
  ).toMatchSnapshot();
});

it('respects first and after and before, too few', () => {
  expect(
    connectionFromArray(letters, {
      first: 2,
      after: offsetToCursor(0),
      before: offsetToCursor(4),
    }),
  ).toMatchSnapshot();
});

it('respects first and after and before, too many', () => {
  expect(
    connectionFromArray(letters, {
      first: 4,
      after: offsetToCursor(0),
      before: offsetToCursor(4),
    }),
  ).toMatchSnapshot();
});

it('respects first and after and before, exactly right', () => {
  expect(
    connectionFromArray(letters, {
      first: 3,
      after: offsetToCursor(0),
      before: offsetToCursor(4),
    }),
  ).toMatchSnapshot();
});

it('respects last and after and before, too few', () => {
  expect(
    connectionFromArray(letters, {
      last: 2,
      after: offsetToCursor(0),
      before: offsetToCursor(4),
    }),
  ).toMatchSnapshot();
});

it('respects last and after and before, too many', () => {
  expect(
    connectionFromArray(letters, {
      last: 4,
      after: offsetToCursor(0),
      before: offsetToCursor(4),
    }),
  ).toMatchSnapshot();
});

it('respects last and after and before, exactly right', () => {
  expect(
    connectionFromArray(letters, {
      last: 3,
      after: offsetToCursor(0),
      before: offsetToCursor(4),
    }),
  ).toMatchSnapshot();
});

it('throws an error if first < 0', () => {
  expect(() => {
    connectionFromArray(letters, { first: -1 });
  }).toThrow('Argument "first" must be a non-negative integer');
});

it('throws an error if last < 0', () => {
  expect(() => {
    connectionFromArray(letters, { last: -1 });
  }).toThrow('Argument "last" must be a non-negative integer');
});

it('returns all elements if cursors are invalid', () => {
  expect(
    connectionFromArray(letters, {
      before: 'invalid',
      after: 'invalid',
    }),
  ).toMatchSnapshot();
});

it('returns all elements if cursors are on the outside', () => {
  expect(
    connectionFromArray(letters, {
      before: offsetToCursor(6),
      after: offsetToCursor(-1),
    }),
  ).toMatchSnapshot();
});

it('returns no elements if cursors cross', () => {
  expect(
    connectionFromArray(letters, {
      before: offsetToCursor(2),
      after: offsetToCursor(4),
    }),
  ).toMatchSnapshot();
});
