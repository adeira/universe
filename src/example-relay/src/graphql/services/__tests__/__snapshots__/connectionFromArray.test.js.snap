// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`respects a smaller first 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": "A",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjE=",
    "hasNextPage": true,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`respects a smaller last 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjQ=",
      "node": "E",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjQ=",
    "hasNextPage": false,
    "hasPreviousPage": true,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjM=",
  },
}
`;

exports[`respects an overly large first 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": "A",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjQ=",
      "node": "E",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjQ=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`respects an overly large last 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": "A",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjQ=",
      "node": "E",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjQ=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`respects first and after 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": true,
    "hasPreviousPage": true,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjI=",
  },
}
`;

exports[`respects first and after and before, exactly right 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": true,
    "hasPreviousPage": true,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjE=",
  },
}
`;

exports[`respects first and after and before, too few 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjI=",
    "hasNextPage": true,
    "hasPreviousPage": true,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjE=",
  },
}
`;

exports[`respects first and after and before, too many 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": true,
    "hasPreviousPage": true,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjE=",
  },
}
`;

exports[`respects first and after with long first 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjQ=",
      "node": "E",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjQ=",
    "hasNextPage": false,
    "hasPreviousPage": true,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjI=",
  },
}
`;

exports[`respects last and after and before, exactly right 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": true,
    "hasPreviousPage": true,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjE=",
  },
}
`;

exports[`respects last and after and before, too few 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": true,
    "hasPreviousPage": true,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjI=",
  },
}
`;

exports[`respects last and after and before, too many 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": true,
    "hasPreviousPage": true,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjE=",
  },
}
`;

exports[`respects last and before 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjI=",
    "hasNextPage": true,
    "hasPreviousPage": true,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjE=",
  },
}
`;

exports[`respects last and before with long last 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": "A",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjI=",
    "hasNextPage": true,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`returns all elements if cursors are invalid 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": "A",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjQ=",
      "node": "E",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjQ=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`returns all elements if cursors are on the outside 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": "A",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjQ=",
      "node": "E",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjQ=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`returns all elements without filters 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": "A",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": "B",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": "C",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": "D",
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjQ=",
      "node": "E",
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjQ=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`returns no elements if cursors cross 1`] = `
Object {
  "edges": Array [],
  "pageInfo": Object {
    "endCursor": null,
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": null,
  },
}
`;
